commandapi-platforms/commandapi-velocity/commandapi-velocity-core/src/main/java/dev/jorel/commandapi/CommandAPIVelocity.java

```diff
@@ -215,8 +215,18 @@ public void postCommandRegistration(RegisteredCommand registeredCommand, Literal
 	}
 
 	@Override
-	public LiteralCommandNode<CommandSource> registerCommandNode(LiteralArgumentBuilder<CommandSource> node) {
-		return getBrigadierDispatcher().register(node);
+	public LiteralCommandNode<CommandSource> registerCommandNode(LiteralArgumentBuilder<CommandSource> node, String namespace) {
+		LiteralCommandNode<CommandSource> builtNode = getBrigadierDispatcher().register(node);
+		if (namespace.isEmpty()) {
+			// We're done. The command already is registered
+			return builtNode;
+		}
+		getBrigadierDispatcher().register(LiteralArgumentBuilder.<CommandSource>literal(namespace + ":" + node.getLiteral())
+			.redirect(builtNode)
+			.requires(node.getRequirement())
+			.executes(builtNode.getCommand())
+		);
```

Comment:
I was thinking about why `CraftServer#syncCommands` on Bukkit copied the command nodes (see [`CommandAPIBukkit#namespaceNode`](https://github.com/JorelAli/CommandAPI/blob/0f8dd2d05e88c168fc1fd7c8894a234ed2cc1fd3/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L528-L543)) rather than using a redirect like we have here. I don't know if they had this problem, but I realized that we probably should be copying the nodes. Consider registering these commands:

```java
new CommandAPICommand("command")
	.withArguments(new LiteralArgument("a"))
	.executes(/* executorA */)
	.register("a");

new CommandAPICommand("command")
	.withArguments(new LiteralArgument("b"))
	.executes(/* executorB */)
	.register("b");
```

Right now, this would create the following command structure:

```
literal `command`:
    literal `a`:
        executes `executorA`
    literal `b`:
        executes `executorB`

literal `a:command`:
    redirects to literal `command`
literal `b:command`:
    redirects to literal `command`
```

The two commands get merged into the `command` literal, which is a little weird, but I don't think that can be avoided. On the other hand, `a:command b` becomes a valid command, even though `a:command` is only supposed to have the `a` literal available. Since the namespaced nodes redirect to the original node, they also get merged together. However, we can avoid this merging by copying the nodes (something similar would need to be implemented in `CommandAPIBukkit` as well), like so:

```java
@Override
public LiteralCommandNode<CommandSource> registerCommandNode(LiteralArgumentBuilder<CommandSource> node, String namespace) {
	LiteralCommandNode<CommandSource> builtNode = getBrigadierDispatcher().register(node);
	if (!namespace.isEmpty()) {
		getBrigadierDispatcher().addChild(namespaceNode(builtNode, namespace));
	}
	return builtNode;
}

private LiteralCommandNode<CommandSource> namespaceNode(LiteralCommandNode<CommandSource> original, String namespace) {
	// Adapted from a section of `org.bukkit.craftbukkit.CraftServer#syncCommands`
	LiteralCommandNode<CommandSource> clone = new LiteralCommandNode<>(
		namespace + ":" + original.getLiteral(),
		original.getCommand(),
		original.getRequirement(),
		original.getRedirect(),
		original.getRedirectModifier(),
		original.isFork()
	);

	for (CommandNode<CommandSource> child : original.getChildren()) {
		clone.addChild(child);
	}

	return clone;
}
```

Then, the example commands from before would look like this:

```
literal `command`:
    literal `a`:
        executes `executorA`
    literal `b`:
        executes `executorB`

literal `a:command`:
    literal `a`:
        executes `executorA`
literal `b:command`:
    literal `b`:
        executes `executorB`
```

The un-namespaced version of the commands still get messed up, but giving the two commands different namespaces allows a namespaced version of the commands to be used as intended without conflict.

As a side note, `CommandAPIBukkit#namespaceNode` doesn't do anything Bukkit-specific; it just needs Brigadier. If we do implement the node-copying solution over here in Velocity, it might make sense to move that method to `CommandAPIHandler` so the logic can be shared.

Reply:
I don't really know what you mean here. Feel free to implement this yourself, I am a little overwhelmed here.

Reply:
When writing tests, I registered this command, following your example command at the top:
```java
CommandAPICommand a = new CommandAPICommand("test")
	.withArguments(LiteralArgument.of("forced"))
	.executesPlayer(info -> {
		results.set("forced");
	});

CommandAPICommand b = new CommandAPICommand("test")
	.withArguments(LiteralArgument.of("notforced"))
	.executesPlayer(info -> {
		results.set("notforced");
	});

a.register("a");
b.register("b");
```

Then, I added tests that should succeed:
```java
server.dispatchCommand(player, "test forced");
assertEquals("forced", results.get());

server.dispatchCommand(player, "a:test forced");
assertEquals("forced", results.get());

server.dispatchCommand(player ,"test notforced");
assertEquals("notforced", results.get());

server.dispatchCommand(player, "b:test notforced");
assertEquals("notforced", results.get());
```

Last, I added tests that should fail:
```java
assertThrows(CommandSyntaxException.class, () -> server.dispatchThrowableCommand(player, "a:test notforced"));
assertThrows(CommandSyntaxException.class, () -> server.dispatchThrowableCommand(player, "b:test forced"));
```
This also succeeded. So I am not really sure if this is a problem _somewhere_ but redirecting seems to work absolutely fine.

Reply:
Ah, I think that happens because the current test is being run after the server is enabled.

For example, this test succeeded:

<details>
<summary>Sucessful test</summary>

```java
@Test
public void registerCommandsWithDifferentNamespacesButSameName() {
	Mut<String> results = Mut.of();
	PlayerMock player = server.addPlayer();

	// Enabling the server before registering commands makes the test succeed
	disablePaperImplementations();
	Bukkit.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
	assertDoesNotThrow(() -> server.getScheduler().performOneTick());
	assertFalse(CommandAPI.canRegister());

	// Test same command name and argument setup but different namespace
	CommandTree a = new CommandTree("test")
		.then(LiteralArgument.of("a").executesPlayer(info -> {
			results.set("a");
		}));

	CommandTree b = new CommandTree("test")
		.then(LiteralArgument.of("b").executesPlayer(info -> {
			results.set("b");
		}));

	a.register("a");
	b.register("b");

	// No namespace command should have been merged
	server.dispatchCommand(player, "test a");
	assertEquals("a", results.get());

	server.dispatchCommand(player, "test b");
	assertEquals("b", results.get());

	// a:test command should be separate from b branch
	server.dispatchCommand(player, "a:test a");
	assertEquals("a", results.get());

	assertThrows(CommandSyntaxException.class, () -> server.dispatchThrowableCommand(player, "a:test b"));

	// b:test command should be separated from a branch
	assertThrows(CommandSyntaxException.class, () -> server.dispatchThrowableCommand(player, "b:test a"));

	server.dispatchCommand(player, "b:test b");
	assertEquals("b", results.get());


	assertNoMoreResults(results);
}
```

</details>

However, this test fails:

<details>
<summary>Failing test</summary>

```java
/* CommandRegistrationTests.registerCommandsWithDifferentNamespacesButSameName:755 
    Expected com.mojang.brigadier.exceptions.CommandSyntaxException to be thrown, 
    but nothing was thrown. */

@Test
public void registerCommandsWithDifferentNamespacesButSameName() {
	Mut<String> results = Mut.of();
	PlayerMock player = server.addPlayer();

	// Test same command name and argument setup but different namespace
	CommandTree a = new CommandTree("test")
		.then(LiteralArgument.of("a").executesPlayer(info -> {
			results.set("a");
		}));

	CommandTree b = new CommandTree("test")
		.then(LiteralArgument.of("b").executesPlayer(info -> {
			results.set("b");
		}));

	a.register("a");
	b.register("b");

	// Enable server after registering commands to call "fixNamespaces"
	disablePaperImplementations();
	Bukkit.getPluginManager().callEvent(new ServerLoadEvent(ServerLoadEvent.LoadType.STARTUP));
	assertDoesNotThrow(() -> server.getScheduler().performOneTick());
	assertFalse(CommandAPI.canRegister());

	// No namespace command should have been merged
	server.dispatchCommand(player, "test a");
	assertEquals("a", results.get());

	server.dispatchCommand(player, "test b");
	assertEquals("b", results.get());

	// a:test command should be separate from b branch
	server.dispatchCommand(player, "a:test a");
	assertEquals("a", results.get());

	// This line (755) fails
	assertThrows(CommandSyntaxException.class, () -> server.dispatchThrowableCommand(player, "a:test b"));

	// b:test command should be separated from a branch
	assertThrows(CommandSyntaxException.class, () -> server.dispatchThrowableCommand(player, "b:test a"));

	server.dispatchCommand(player, "b:test b");
	assertEquals("b", results.get());


	assertNoMoreResults(results);
}
```

</details>

Whether or not the server is enabled matters because the two cases are handled by different code.

When the server is enabled, this code creates the namespaced version:

https://github.com/JorelAli/CommandAPI/blob/1952d332627ee2c04854c41b19997746c6e5cf3e/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L480-L549

This path uses `CommandAPIBukkit#namespaceNode` to copy the original node as a namespaced version.

When the server isn't enabled, it uses this code to create the namespaced version:

https://github.com/JorelAli/CommandAPI/blob/1952d332627ee2c04854c41b19997746c6e5cf3e/commandapi-platforms/commandapi-bukkit/commandapi-bukkit-core/src/main/java/dev/jorel/commandapi/CommandAPIBukkit.java#L551-L572

This uses the redirect method. 

So, when the server is enabled, `a:test` is a clone of the `a` command, and it correctly can not access the `b` branch as you saw. However, if the server is not enabled, `a:test` is redirected to `test`, so running `a:test b` incorrectly succeeds, and the test fails. To solve this, we should probably always use `CommandAPIBukkit#namespaceNode` to copy the node instead of redirecting, like so:

```java
@Override
public LiteralCommandNode<Source> registerCommandNode(LiteralArgumentBuilder<Source> node, String namespace) {
	LiteralCommandNode<Source> builtNode = getBrigadierDispatcher().register(node);
	if (namespace.isEmpty()) {
		// Bukkit will automatically create `minecraft:command`
		// We want to remove that so there is no namespace
		namespacesToFix.add("minecraft:" + node.getLiteral());
	} else if(!namespace.equals("minecraft")) {
		// If Bukkit won't create our namespaces, we'll do it ourselves
		getBrigadierDispatcher().getRoot().addChild(namespaceNode(builtNode, namespace));

		// Make sure Bukkit's automatic `minecraft:` namespaces don't stay
		String name = node.getLiteral();
		String namespacedName = namespace + ":" + name;
		namespacesToFix.add("minecraft:" + name);
		namespacesToFix.add("minecraft:" + namespacedName);
	}
	return builtNode;
}
```

In general, since there are two separate code paths in action, we should probably ensure all the tests work the same when registering the commands before and after the server is enabled to catch cases like this. I also noticed that `CommandRegistrationTests` is pretty big, so it might be a good idea to separate the namespace tests into separate methods in another class called `CommandNamespaceTests` or something.